### PR TITLE
Support session_vars and client_params in S3()

### PR DESCRIPTION
**Pull request by romain-intel**
This allows the user to pass down session variables and client parameters
to set when getting a S3 client.

### CODE REVIEW

File: metaflow/metaflow_config.py [hunk 1]
```diff
@@ -68,6 +102,31 @@ def from_conf(name, default=None):
...
     else None,
 )
+
+DATATOOLS_DEFAULT_CLIENT_PARAMS = from_conf(
+    "METAFLOW_DATATOOLS_CLIENT_PARAMS",
+    '{"endpoint_url": %s, "verify": %s}' % (S3_ENDPOINT_URL, S3_VERIFY_CERTIFICATE)
+    if S3_ENDPOINT_URL and S3_VERIFY_CERTIFICATE
+    else "{}",
+)
+DATATOOLS_DEFAULT_SESSION_VARS = from_conf("METAFLOW_DATATOOLS_SESSION_VARS", "{}")
```
Comment (oavdeev) [inline metaflow/metaflow_config.py:112]: nit^2: maybe move this line closer to where the value is parsed at line 122. At the first glance I thought it was a third configuration variable.

File: metaflow/metaflow_config.py [hunk 2]
```diff
@@ -68,6 +102,31 @@ def from_conf(name, default=None):
     if from_conf("METAFLOW_DATASTORE_SYSROOT_S3")
     else None,
 )
+
+DATATOOLS_DEFAULT_CLIENT_PARAMS = from_conf(
+    "METAFLOW_DATATOOLS_CLIENT_PARAMS",
+    '{"endpoint_url": %s, "verify": %s}' % (S3_ENDPOINT_URL, S3_VERIFY_CERTIFICATE)
```
Comment (oavdeev) [inline metaflow/metaflow_config.py:108]: I think you'd want to to quote `endpoint_url`. Also there gotta be a more robust way to construct that default value without raw string formatting, idk `json.dumps()` maybe

File: metaflow/metaflow_config.py [hunk 3]
```diff
@@ -68,6 +102,31 @@ def from_conf(name, default=None):
...
+    if S3_ENDPOINT_URL and S3_VERIFY_CERTIFICATE
```
[the remark this answers is not in the document]
Comment (oavdeev) [reply metaflow/metaflow_config.py:109]: Overall it is kind of hard to follow the logic for this one. Sounds like setting `S3_ENDPOINT_URL` has no effect if you also set `DATATOOLS_DEFAULT_CLIENT_PARAMS` ?

Can you change the code or at least add a comment how all these options interact?

**Review by oavdeev — CHANGES_REQUESTED** [review]

File: metaflow/metaflow_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (romain-intel) [reply metaflow/metaflow_config.py:109]: sure. And yes, the idea was to be as backwards compatible as possible. METAFLOW_DATATOOLS_CLIENT_PARAMS is new so if you don't use that you can continue to use both S3_ENDPOINT_URL and/or S3_VERIFY_CERTIFICATE. I can change the condition to an `or` if you want. That's a good call out actually; I hadn't realized that.

File: metaflow/metaflow_config.py (hunk 2 above)
Comment (romain-intel) [reply metaflow/metaflow_config.py:108]: or rather: 
```
json.dumps({"endpoint_url": S3_ENDPOINT_URL, "verify": S3_VERIFY_CERTIFICATE})
```

File: metaflow/metaflow_config.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (oavdeev) [reply metaflow/metaflow_config.py:109]: Yeah I'm not sure what's the best logic, just wanted to make it clear to the reader :) Btw could also _merge_ `endpoint_url` into  `METAFLOW_DATATOOLS_CLIENT_PARAMS` dictionary if both are specified.

**Review by oavdeev — APPROVED** [review]
Lgtm!

**Review by savingoyal — APPROVED** [review]
